src/analysis/convex/cone/proper.lean

```diff
@@ -46,8 +45,106 @@ protected def closure (K : convex_cone ℝ E) : convex_cone ℝ E :=
 
 @[simp, norm_cast] lemma coe_closure (K : convex_cone ℝ E) : (K.closure : set E) = closure K := rfl
 
-protected lemma mem_closure {K : convex_cone ℝ E} {a : E} :
-  a ∈ K.closure ↔ a ∈ closure (K : set E) :=
-iff.rfl
+@[simp] protected lemma mem_closure {K : convex_cone ℝ E} {a : E} :
+  a ∈ K.closure ↔ a ∈ closure (K : set E) := iff.rfl
+
+lemma closure_eq_iff_is_closed {K : convex_cone ℝ E} : K.closure = K ↔ is_closed (K : set E) :=
+⟨ (λ h, by rw [← closure_eq_iff_is_closed, ← coe_closure, h]),
+  (λ h, set_like.coe_injective $ closure_eq_iff_is_closed.2 h) ⟩
 
 end convex_cone
+
+section definitions
+
+/-- A proper cone is a convex cone `K` that is nonempty and closed. Proper cones have the nice
+property that the dual of the dual of a proper cone is itself. This makes them useful for defining
+cone programs and proving duality theorems. -/
+structure proper_cone (E : Type*) [normed_add_comm_group E] [inner_product_space ℝ E]
+  extends convex_cone ℝ E :=
+(nonempty'  : (carrier : set E).nonempty)
+(is_closed' : is_closed (carrier : set E))
+
+end definitions
+
+namespace proper_cone
+
+section inner_product_space
+
+variables {E : Type*} [normed_add_comm_group E] [inner_product_space ℝ E]
+variables {F : Type*} [normed_add_comm_group F] [inner_product_space ℝ F]
+
+instance : has_coe (proper_cone E) (convex_cone ℝ E) := ⟨λ K, K.1⟩
+
+instance : has_mem E (proper_cone E) := ⟨λ e K, e ∈ (K : convex_cone ℝ E) ⟩
+
+noncomputable instance : has_zero (proper_cone E) :=
+⟨ { to_convex_cone := (0 : convex_cone ℝ E),
+    nonempty' := ⟨0, rfl⟩,
+    is_closed' := is_closed_singleton } ⟩
+
+noncomputable instance : inhabited (proper_cone E) := ⟨0⟩
+
+@[simp] lemma mem_coe {x : E} {K : proper_cone E} : x ∈ (K : convex_cone ℝ E) ↔ x ∈ K := iff.rfl
+
+protected lemma nonempty (K : proper_cone E) : (K : set E).nonempty := K.nonempty'
+
+protected lemma is_closed (K : proper_cone E) : is_closed (K : set E) := K.is_closed'
+
+protected lemma pointed (K : proper_cone E) : (K : convex_cone ℝ E).pointed :=
+(K : convex_cone ℝ E).pointed_of_nonempty_of_is_closed K.nonempty K.is_closed
+
+lemma ext' {S T : proper_cone E} (h : (S : convex_cone ℝ E) = T) : S = T :=
+by cases S; cases T; congr'
+
+instance : set_like (proper_cone E) E :=
+{ coe := λ K, K.carrier,
+  coe_injective' := λ _ _ h, proper_cone.ext' (set_like.coe_injective h) }
```

Comment:
Doesn't `set_like` give you automatically a `has_mem` instance? I.e., can you try to remove the `has_mem` instance and see whether `x ∈ S` for `S : proper_cone E` still type-checks? If it doesn't it might be better to use the definition from `set_like`.

Reply:
It does work. The `has_mem` for `set_like` is sufficient. I got rid of the redundant lemma.